src/dataflow/src/source/loki.rs

```diff
@@ -21,48 +32,94 @@ struct ConnectionInfo {
-                .query(&[("direction", "forward")])
-                .send()
-                .await
+        client
+            .get(format!("{}/loki/api/v1/query_range", conn_info.endpoint))
+            .basic_auth(conn_info.user, Some(conn_info.pw))
+            .query(&[("query", query)])
+            .query(&[("start", format!("{}", start))])
+            .query(&[("end", format!("{}", end))])
+            .query(&[("direction", "forward")])
+            .send()
+            .await
     }
 
-    fn new_stream(self) -> impl stream::Stream<Item = Result<QueryResult, reqwest::Error>> {
-        let polls = IntervalStream::new(tokio::time::interval(self.batch_window));
+    fn new_stream(
+        query: String,
+        batch_window: Duration,
+        conn_info: ConnectionInfo,
+    ) -> impl stream::Stream<Item = Result<QueryResult, reqwest::Error>> {
+        let polls = IntervalStream::new(tokio::time::interval(batch_window));
 
-        let batch_window = self.batch_window;
-        let conn_info = self.conn_info.clone();
+        let conn_info = conn_info;
         polls.then(move |_tick| {
             let start = SystemTime::now() - batch_window;
             let end = SystemTime::now();
 
             Self::query(
                 conn_info.clone(),
                 start.duration_since(UNIX_EPOCH).unwrap().as_nanos(),
-                end.duration_since(UNIX_EPOCH).unwrap().as_nanos())
-                .and_then(|resp| { resp.json::<QueryResult>() })
+                end.duration_since(UNIX_EPOCH).unwrap().as_nanos(),
+                query.clone(),
+            )
+            .and_then(|resp| resp.json::<QueryResult>())
         })
     }
 }
 
 #[async_trait]
 impl SimpleSource for LokiSourceReader {
-
     async fn start(mut self, timestamper: &Timestamper) -> Result<(), SourceError> {
-        let stream = self.new_stream();
+        let stream = Self::new_stream(self.query, self.batch_window, self.conn_info);
+        tokio::pin!(stream);
+
+        while let Some(entry) = stream.next().await {
+            match entry {
+                Ok(result) => {
+                    let Data::Streams(streams) = result.data;
+
+                    let lines: Vec<String> = streams
+                        .iter()
+                        .flat_map(|s| {
+                            s.values.iter().map(|v| {
+                                json!({"line": v.line.clone(), "labels": s.labels}).to_string()
+                            })
+                        })
+                        .collect();
+
+                    for line in lines {
+                        let row = Row::pack_slice(&[Datum::String(&line)]);
+
+                        timestamper.insert(row).await.map_err(|e| SourceError {
+                            source_id: self.source_id.clone(),
+                            error: SourceErrorDetails::FileIO(e.to_string()),
+                        })?;
+                    }
```

Comment:
We should probably be adding all of these lines at the same timestamp (i.e. as part of the same `Row`, or by using [`Timestamper::start_tx`](https://dev.materialize.com/api/rust/mz_dataflow/source/struct.Timestamper.html#method.start_tx)). Might be able to do something like...

```suggestion
                    let rows: Vec<_> = lines
                        .iter()
                        .map(|l| Datum::String(&line))
                        .collect();
                    let row = Row::pack_slice(&rows);
                    timestamper.insert(row).await.map_err(|e| SourceError {
                        source_id: self.source_id.clone(),
                        error: SourceErrorDetails::FileIO(e.to_string()),
                    })?;
```

Reply:
Hm, I think this would add everything as one row.